src/main/java/com/blamejared/crafttweaker/impl/recipes/Replacer.java

```diff
@@ -157,31 +165,223 @@ public static Replacer forAllTypesExcluding(final IRecipeManager... managers) {
+        this.userRenames.put(oldName, actualNewName);
+        return this;
+    }
+    
+    /**
+     * Specifies the {@link BiFunction} that will be used for renaming all recipes.
+     *
+     * <p>The first argument to the function is the {@link ResourceLocation} that uniquely identifies its name, whereas
+     * the second represents the default name for the recipe according to the default replacement rules. The return
+     * value of the function will then represent the new name of the recipe.</p>
+     *
+     * @param function The renaming function.
+     * @return This replacer for chaining
+     *
+     * @docParam function myFunction
+     */
+    @ZenCodeType.Method
+    public Replacer useForRenaming(final BiFunction<ResourceLocation, String, String> function) {
+        if (this.userRenamingFunction != null) {
+            final CodePosition position = PositionUtil.getZCScriptPositionFromStackTrace();
+            CraftTweakerAPI.logWarning("%sA renaming function has already been specified for this replacer: the old one will be replaced", position == null? "" : position + ": ");
+        }
+        
+        this.userRenamingFunction = function;
+        return this;
+    }
+    
+    /**
+     * Executes all replacements that have been queued on this replacer, if any.
+     */
     @ZenCodeType.Method
     public void execute() {
         if (this.rules.isEmpty()) return;
-        CraftTweakerAPI.apply(new ReplacerAction(this.targetedManagers, this.targetedRecipes, Collections.unmodifiableList(this.rules)));
+        CraftTweakerAPI.apply(
+                new ReplacerAction(
+                        this.targetedManagers,
+                        this.targetedRecipes,
+                        Collections.unmodifiableList(this.rules),
+                        this.targetedManagers.stream().flatMap(manager -> DEFAULT_EXCLUSIONS.computeIfAbsent(manager, this::gatherDefaultExclusions).stream()).collect(Collectors.toList()),
+                        Collections.unmodifiableCollection(this.userExclusionList),
+                        this.buildGeneratorFunction()
+                )
+        );
     }
     
-    private Replacer addReplacementRule(final IReplacementRule rule) {
+    // Keep public but not exposed to Zen: this is public API (yeah, I know, bad placement)
+    public Replacer addReplacementRule(final IReplacementRule rule) {
         if (rule == IReplacementRule.EMPTY) return this;
         this.rules.add(rule);
         return this;
     }
+    
+    private String fix(final String newName, final ResourceLocation oldName) {
+        final CodePosition position = PositionUtil.getZCScriptPositionFromStackTrace();
+        return NameUtils.fixing(
+                newName,
+                (fixed, mistakes) -> CraftTweakerAPI.logWarning(
+                        "%sInvalid recipe rename '%s' from '%s', mistakes:\n%s\nThe new rename '%s' will be used",
+                        position == null ? "" : position + ": ",
```

Comment:
Check the implementation of `getZCScriptPositionFromStackTrace` again, but I think it never returned null but rather `CodePosition.UNKNOWN`